codec-http2/src/main/java/io/netty/handler/codec/http2/UniformStreamByteDistributor.java

```diff
@@ -0,0 +1,252 @@
+        int allocationChunk = max(minAllocationChunk, (int)(unallocated / queue.size()));
+        while (maxBytes > 0 && !queue.isEmpty()) {
+            //int allocationChunk = max(minAllocationChunk, (int)(unallocated / queue.size()));
+
+            // Remove the head of the queue.
+            State state = queue.remove();
+            state.enqueued = false;
+
+            // Allocate as much data as we can.
+            int chunk = min(allocationChunk, min(maxBytes, state.unallocated()));
+            state.allocate(chunk);
+            maxBytes -= chunk;
+            allocated += chunk;
+            unallocated -= chunk;
+
+            if (state.unallocated() > 0) {
+                // The stream still has unallocated bytes, add it back to the end of the queue.
+                state.enqueue();
+            }
+        }
+
+        return allocated;
+    }
+
+    private State state(Http2Stream stream) {
+        return checkNotNull(stream, "stream").getProperty(stateKey);
+    }
+
+    private class EnqueueStrategy {
+        void enqueue(State state) {
+            queue.addLast(state);
+            state.enqueued = true;
+        }
+    }
+
+    private final EnqueueStrategy defaultEnqueueStrategy = new EnqueueStrategy();
+
+    /**
+     * The first time a state is enqueued, it will be added to the head of the queue. After
+     * that it will be added to the tail.
+     */
+    private final EnqueueStrategy firstEnqueueStrategy = new EnqueueStrategy() {
+        @Override
+        void enqueue(State state) {
+            queue.addFirst(state);
+            state.enqueued = true;
+
+            // Switch to the default strategy so that it will be added to the tail from
+            // now on.
+            state.enqueueStrategy = defaultEnqueueStrategy;
+        }
+    };
+
+    /**
+     * The remote flow control state for a single stream.
+     */
+    private final class State {
+        final Http2Stream stream;
+        boolean hasFrame;
+        int streamableBytes;
+        int allocated;
+        boolean enqueued;
+        EnqueueStrategy enqueueStrategy = firstEnqueueStrategy;
+
+        State(Http2Stream stream) {
+            this.stream = stream;
+        }
+
+        void allocate(int bytes) {
+            allocated += bytes;
+        }
+
+        int unallocated() {
+            return streamableBytes - allocated;
+        }
+
+        /**
+         * Reset the number of bytes that have been allocated to this stream by the priority
+         * algorithm.
+         */
+        void resetAllocated() {
+            allocated = 0;
+        }
+
+        void updateStreamableBytes(int newStreamableBytes, boolean hasFrame) {
+            this.hasFrame = hasFrame;
+
+            int delta = newStreamableBytes - streamableBytes;
+            if (delta != 0) {
+                streamableBytes = newStreamableBytes;
+                totalStreamableBytes += delta;
+
+                if (!enqueued && streamableBytes > 0) {
```

Comment:
I see. Same as the priority distributor you are always allocating over all active streams for the write.  I guess we could potentially limit this to what is in the queue if we include EMPTY elements ... but doesn't seem necessary now.